pandas/_libs/groupby.pyx

```diff
@@ -382,5 +382,61 @@ def group_any_all(uint8_t[:] out,
                 out[lab] = flag_val
 
 
+ctypedef fused floating:
+    float32_t
```

Comment:
we likely have this type elsewhere

Reply:
cython has a builtin `floating` fused type for float+double, better to re-use that as long as we're OK with declaring that float==float32_t and double==float64_t

Reply:
@jbrockmendel I made a change that the group_add function is not looked up by type (e.g. grou_add_float64).

It seems that after the change, when group_add is used with a ndarray[int64], the ndarray is first cast into ndarray[float64], then passed into group_add[float64], then cast back into ndarray[int64].

This is different then before the change (which passed the tests), because the cast back into ndarray[int64] did not occur so the result was an ndarray[float64].

See [here](https://github.com/pandas-dev/pandas/blob/master/pandas/core/groupby/ops.py#L476).

What do you think is the best solution? keep the group_add_{type} = group_add_floating statements?.

Reply:
> [...] ndarray is first cast into ndarray[float64] [...]

This is definitely not desirable.  Any guesses why this might be happening?

> What do you think is the best solution? keep the group_add_{type} = group_add_floating statements?

The best case would be to just have `group_add` and when we call it from outside cython, cython will figure out what its supposed to do (correctly, without doing an unnecessary casting).  In my experience, doing this lets us get rid of a bunch of ugly non-cython code.

Next best would be to keep `group_add_dtype` functions.  In these cases, I wouldn't suggest `group_add_{type} = group_add_floating` but more specifically `group_add_{type} = group_add_floating[{type}]`.  And if feasible, I'd prefer `group_add[{type}]` over `group_add_floating[{type}]`.

Is this helpful?

Reply:
If you change the `floating` types to `numeric` does that preserve it?

Reply:
@WillAyd the function recieves a 2d array (out) which it fills with either the actual result, or NaN in some cases.
AFAIK NaN in numpy is a special value for floats and does not exist for integers.

Reply:
@jbrockmendel 
For some reason I could only make the specialization indexing work with strings rather than the types (e.g. group_add_float32 = group_add_floating['float32_t'].

I suspect it might have something to do with trying to store a reference to the function rather than calling it, not sure.

I've looked into keeping the name group_add without the group_add_{type} members (i.e. changing things at the caller side).
In groupby/ops.py the method _get_cython_function can be changed to support limiting types for functions. The function are taken from _cython_functions which already supports dictionaries with the following keys:

* name - the name of the function
* f - lambda/method that wraps the function

a third key, supported types, could solve this issue. When the actual type is not in the supported type None would be returned. This way the group_add_{type} = ... statements could be removed.

Which solution do you think is best?

Reply:
@noamher generally we want to try and simplify the calls we are making from the groupby module back to the Cython layer, so I'd be a little hesitant to add more logic therein.

Might be missing the point but can you provide a specific example of the issue you are trying to solve? Certain types simply not resolving to the function correctly?

Reply:
Just to be clear, the specialization works. I preferred to use the actual types rather than string thouh but couldn't make it work.

I guess the cyhton layer can do the cast into float and back into int.
I'll try to understand what happens to the nans when the output is cast back into floats.

Reply:
> I preferred to use the actual types rather than string thouh but couldn't make it work.

Yah I've found this tricky too.  Don't sweat it.

> I've looked into keeping the name group_add without the group_add_{type} members (i.e. changing things at the caller side).

Agreed with @WillAyd, definitely don't want to add more complexity on the python side.  Removing the `group_add_{type}`s would be nice, but if we have to do that in a follow-up that's OK too.

Reply:
Used cython.floating instead of creating a fused type, is that okay?